File: web/src/model/victron_controller/dashboard/ActuatedF64.ts
// @ts-nocheck
import {BaboonGeneratedLatest, BaboonCodecContext, BaboonBinWriter, BinTools, BaboonBinReader, Lazy} from '../../BaboonSharedRuntime'
import {TargetPhase, TargetPhase_UEBACodec} from './TargetPhase'
import {Owner, Owner_UEBACodec} from './Owner'
import {ActualF64, ActualF64_UEBACodec} from './ActualF64'

export class ActuatedF64 implements BaboonGeneratedLatest {
    private readonly _target_value: number | undefined;
    private readonly _target_owner: Owner;
    private readonly _target_phase: TargetPhase;
    private readonly _target_since_epoch_ms: bigint;
    private readonly _actual: ActualF64;

    constructor(target_value: number | undefined, target_owner: Owner, target_phase: TargetPhase, target_since_epoch_ms: bigint, actual: ActualF64) {
        this._target_value = target_value
        this._target_owner = target_owner
        this._target_phase = target_phase
        this._target_since_epoch_ms = target_since_epoch_ms
        this._actual = actual
    }

    public get target_value(): number | undefined {
        return this._target_value;
    }
    public get target_owner(): Owner {
        return this._target_owner;
    }
    public get target_phase(): TargetPhase {
        return this._target_phase;
    }
    public get target_since_epoch_ms(): bigint {
        return this._target_since_epoch_ms;
    }
    public get actual(): ActualF64 {
        return this._actual;
    }

    public toJSON(): Record<string, unknown> {
        return {
            target_value: this._target_value !== undefined ? this._target_value : undefined,
            target_owner: this._target_owner,
            target_phase: this._target_phase,
            target_since_epoch_ms: this._target_since_epoch_ms,
            actual: this._actual
        };
    }

    public with(overrides: {target_value?: number | undefined; target_owner?: Owner; target_phase?: TargetPhase; target_since_epoch_ms?: bigint; actual?: ActualF64}): ActuatedF64 {
        return new ActuatedF64(
            'target_value' in overrides ? overrides.target_value! : this._target_value,
            'target_owner' in overrides ? overrides.target_owner! : this._target_owner,
            'target_phase' in overrides ? overrides.target_phase! : this._target_phase,
            'target_since_epoch_ms' in overrides ? overrides.target_since_epoch_ms! : this._target_since_epoch_ms,
            'actual' in overrides ? overrides.actual! : this._actual
        );
    }

    public static fromPlain(obj: {target_value: number | undefined; target_owner: Owner; target_phase: TargetPhase; target_since_epoch_ms: bigint; actual: ActualF64}): ActuatedF64 {
        return new ActuatedF64(
            obj.target_value,
            obj.target_owner,
            obj.target_phase,
            obj.target_since_epoch_ms,
            obj.actual
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return ActuatedF64.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return ActuatedF64.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#ActuatedF64'
    public baboonTypeIdentifier() {
        return ActuatedF64.BaboonTypeIdentifier
    }
    public static readonly BaboonSameInVersions = ["0.1.0", "0.2.0", "0.3.0"]
    public baboonSameInVersions() {
        return ActuatedF64.BaboonSameInVersions
    }
    public static binCodec(): ActuatedF64_UEBACodec {
        return ActuatedF64_UEBACodec.instance
    }
}

export class ActuatedF64_UEBACodec {
    public encode(ctx: BaboonCodecContext, value: ActuatedF64, writer: BaboonBinWriter): unknown {
        if (this !== ActuatedF64_UEBACodec.lazyInstance.value) {
          return ActuatedF64_UEBACodec.lazyInstance.value.encode(ctx, value, writer)
        }
    
        if (ctx === BaboonCodecContext.Indexed) {
            BinTools.writeByte(writer, 0x01);
            const buffer = new BaboonBinWriter();
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                if (value.target_value === undefined) {
                BinTools.writeByte(buffer, 0);
            } else {
                BinTools.writeByte(buffer, 1);
                BinTools.writeF64(buffer, value.target_value);
            }
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            Owner_UEBACodec.instance.encode(ctx, value.target_owner, buffer);
            TargetPhase_UEBACodec.instance.encode(ctx, value.target_phase, buffer);
            BinTools.writeI64(buffer, value.target_since_epoch_ms);
            {
                const before = buffer.position();
                BinTools.writeI32(writer, before);
                ActualF64_UEBACodec.instance.encode(ctx, value.actual, buffer);
                const after = buffer.position();
                BinTools.writeI32(writer, after - before);
            }
            writer.writeAll(buffer.toBytes());
        } else {
            BinTools.writeByte(writer, 0x00)
            if (value.target_value === undefined) {
                BinTools.writeByte(writer, 0);
            } else {
                BinTools.writeByte(writer, 1);
                BinTools.writeF64(writer, value.target_value);
            }
            Owner_UEBACodec.instance.encode(ctx, value.target_owner, writer);
            TargetPhase_UEBACodec.instance.encode(ctx, value.target_phase, writer);
            BinTools.writeI64(writer, value.target_since_epoch_ms);
            ActualF64_UEBACodec.instance.encode(ctx, value.actual, writer);
        }
    }
    
    public decode(ctx: BaboonCodecContext, reader: BaboonBinReader): ActuatedF64 {
        if (this !== ActuatedF64_UEBACodec .lazyInstance.value) {
            return ActuatedF64_UEBACodec.lazyInstance.value.decode(ctx, reader)
        }
    
        const header = BinTools.readByte(reader);
        const useIndices = header === 0x01;
        if (useIndices) {
            for (let i = 0; i < 2; i++) {
                BinTools.readI32(reader);
                BinTools.readI32(reader);
            }
        }
        const target_value = (BinTools.readByte(reader) === 0 ? undefined : BinTools.readF64(reader));
        const target_owner = Owner_UEBACodec.instance.decode(ctx, reader);
        const target_phase = TargetPhase_UEBACodec.instance.decode(ctx, reader);
        const target_since_epoch_ms = BinTools.readI64(reader);
        const actual = ActualF64_UEBACodec.instance.decode(ctx, reader);
        return new ActuatedF64(
            target_value,
            target_owner,
            target_phase,
            target_since_epoch_ms,
            actual,
        );
    }

    public static readonly BaboonDomainVersion = '0.3.0'
    public baboonDomainVersion() {
        return ActuatedF64_UEBACodec.BaboonDomainVersion
    }
    public static readonly BaboonDomainIdentifier = 'victron_controller.dashboard'
    public baboonDomainIdentifier() {
        return ActuatedF64_UEBACodec.BaboonDomainIdentifier
    }
    public static readonly BaboonTypeIdentifier = 'victron_controller.dashboard/:#ActuatedF64'
    public baboonTypeIdentifier() {
        return ActuatedF64_UEBACodec.BaboonTypeIdentifier
    }

    protected static lazyInstance = new Lazy(() => new ActuatedF64_UEBACodec())
    public static get instance(): ActuatedF64_UEBACodec {
        return ActuatedF64_UEBACodec.lazyInstance.value
    }
}

File: crates/dashboard-model/src/victron_controller/dashboard/v0_2_0/actuated_f64.rs
use crate::victron_controller::dashboard::v0_2_0::actual_f64::ActualF64;
use crate::victron_controller::dashboard::v0_2_0::owner::Owner;
use crate::victron_controller::dashboard::v0_2_0::target_phase::TargetPhase;

#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
pub struct ActuatedF64 {
    pub target_value: Option<f64>,
    pub target_owner: Owner,
    pub target_phase: TargetPhase,
    #[serde(deserialize_with = "crate::baboon_runtime::lenient_numeric::deserialize")]
    pub target_since_epoch_ms: i64,
    pub actual: ActualF64,
}

impl PartialEq for ActuatedF64 {
    fn eq(&self, other: &Self) -> bool {
        self.cmp(other) == std::cmp::Ordering::Equal
    }
}

impl Eq for ActuatedF64 {}

impl PartialOrd for ActuatedF64 {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for ActuatedF64 {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        match crate::baboon_runtime::__opt_f64_total_cmp(&self.target_value, &other.target_value) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.target_owner.cmp(&other.target_owner) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.target_phase.cmp(&other.target_phase) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.target_since_epoch_ms.cmp(&other.target_since_epoch_ms) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        match self.actual.cmp(&other.actual) {
            std::cmp::Ordering::Equal => {},
            ord => return ord,
        }
        std::cmp::Ordering::Equal
    }
}

impl crate::baboon_runtime::BaboonBinCodecIndexed for ActuatedF64 {
    fn index_elements_count(_ctx: &crate::baboon_runtime::BaboonCodecContext) -> u16 {
        2
    }
}

impl crate::baboon_runtime::BaboonBinEncode for ActuatedF64 {
    fn encode_ueba(&self, ctx: &crate::baboon_runtime::BaboonCodecContext, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let value = self;
        if ctx.use_indices() {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x01)?;
            let mut buffer: Vec<u8> = Vec::new();
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                match &value.target_value {
                None => crate::baboon_runtime::bin_tools::write_byte(&mut buffer, 0)?,
                Some(v) => {
                    crate::baboon_runtime::bin_tools::write_byte(&mut buffer, 1)?;
                    v.encode_ueba(ctx, &mut buffer)?;
                }
            }
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            value.target_owner.encode_ueba(ctx, &mut buffer)?;
            value.target_phase.encode_ueba(ctx, &mut buffer)?;
            value.target_since_epoch_ms.encode_ueba(ctx, &mut buffer)?;
            {
                let before = buffer.len();
                crate::baboon_runtime::bin_tools::write_i32(writer, before as i32)?;
                value.actual.encode_ueba(ctx, &mut buffer)?;
                let after = buffer.len();
                let length = after - before;
                crate::baboon_runtime::bin_tools::write_i32(writer, length as i32)?;
            }
            writer.write_all(&buffer)?;
        } else {
            crate::baboon_runtime::bin_tools::write_byte(writer, 0x00)?;
            match &value.target_value {
                None => crate::baboon_runtime::bin_tools::write_byte(writer, 0)?,
                Some(v) => {
                    crate::baboon_runtime::bin_tools::write_byte(writer, 1)?;
                    v.encode_ueba(ctx, writer)?;
                }
            }
            value.target_owner.encode_ueba(ctx, writer)?;
            value.target_phase.encode_ueba(ctx, writer)?;
            value.target_since_epoch_ms.encode_ueba(ctx, writer)?;
            value.actual.encode_ueba(ctx, writer)?;
        }
        Ok(())
    }
}

impl crate::baboon_runtime::BaboonBinDecode for ActuatedF64 {
    fn decode_ueba(ctx: &crate::baboon_runtime::BaboonCodecContext, reader: &mut dyn std::io::Read) -> Result<Self, Box<dyn std::error::Error>> {
        let (_header, index) = <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::read_index(ctx, reader)?;
        if ctx.use_indices() {
            assert_eq!(index.len(), <Self as crate::baboon_runtime::BaboonBinCodecIndexed>::index_elements_count(ctx) as usize);
        }
        let target_value = {
            let tag = crate::baboon_runtime::bin_tools::read_byte(reader)?;
            if tag == 0 { None } else { Some(crate::baboon_runtime::bin_tools::read_f64(reader)?) }
        };
        let target_owner = Owner::decode_ueba(ctx, reader)?;
        let target_phase = TargetPhase::decode_ueba(ctx, reader)?;
        let target_since_epoch_ms = crate::baboon_runtime::bin_tools::read_i64(reader)?;
        let actual = ActualF64::decode_ueba(ctx, reader)?;
        Ok(ActuatedF64 {
            target_value,
            target_owner,
            target_phase,
            target_since_epoch_ms,
            actual,
        })
    }
}